Manage api from Cachet Settings page (#235)

icecoldPHP opened the pull request:
Add settings to enable and disable API
Add settings to force authentication for API
Closes: #235

icecoldPHP committed:
Add settings to enable and disable API
Add settings to force authentication for API

icecoldPHP committed:
Small typo

icecoldPHP committed:
Composer fail fix

icecoldPHP committed:
Composer fail fix

icecoldPHP committed:
Hide api protected when api is disabled

jbrooksuk commented on src/CachetCoreServiceProvider.php:
```diff
@@ -111,14 +112,36 @@ private function configureRateLimiting(): void
+                if(! $application->runningInConsole()) {
+                    throw new \Exception("Please run `php artisan migrate` to load missing settings.");
+                }
+            } catch (\Exception $exception) {
+                // do nothing
+            }
```
This can be removed if it's not doing anything.

icecoldPHP replied:
The catch has to be there.
for the sole purpose that when there is no DB or table the application still works:
https://github.com/cachethq/core/pull/236/checks?check_run_id=36911121166

joelbutcher replied:
@icecoldPHP in what scenario will the database and / or table(s) not exist for this feature?

icecoldPHP replied:
@joelbutcher If you look at the check.
When you setup cachet for the first time there is no DB.
since this code is in the Service Provider it is always called. and since we're using Spatie Settings.
It's trying to look for the data in the DB which there isn't. So basicly it's failing as long as there is no Table to read from. Secondly is where Unit Testing comes in to play where i'm currently also running into the same issue.
Since it's also giving me the same error that the DB is not initialized

jbrooksuk commented on src/CachetCoreServiceProvider.php:
```diff
@@ -111,14 +112,36 @@ private function configureRateLimiting(): void
+            if ($apiEnabled) {
+                $middleware = ['cachet:api', 'throttle:cachet-api'];
+                if ($apiProtected) {
+                    $middleware[] = 'auth:sanctum';
```
This middleware is already added to the non-GET routes. I think we should consider adding a middleware that is always applied, but conditionally authorises if the setting is enabled.

jbrooksuk reviewed (changes requested):
Overall this looks pretty good to me. We need some tests to validate that the API can be enabled and disabled + that `GET` requests can be protected.

icecoldPHP committed:
Moved api check to 2 different middleware.
ApiEnabled & ApiPublicOrProtected which solves the early access to DB
Added GetOrDefault to AppSettings for ease of use and removal of duplicate code.

icecoldPHP committed:
Api Enabled should not have to be extended from Auth

jbrooksuk committed:
Refactor

jbrooksuk commented on src/Http/Middleware/ApiPublicOrProtected.php:
```diff
@@ -0,0 +1,31 @@
+    {
+        $protected = AppSettings::getOrDefault('api_protected', false);
+        if ($protected) {
+            return parent::handle($request, $next, ...$guards);
```
This is trying to redirect to a `login` route which doesn't exist in Cachet.

icecoldPHP replied:
Will do

icecoldPHP replied:
@jbrooksuk This is something that is managed by the Sanctum API.
We could override this because this is currently a global issue.
The same is for other things that are broken between Core and Cachet repo's.
Because in the Cachet Repo the user model is missing fields so it cannot set the admin permissions.
When you say yes in the terminal.

jbrooksuk commented on src/Settings/AppSettings.php:
```diff
@@ -32,8 +33,33 @@ class AppSettings extends Settings
+     * @return mixed
+     * @throws \Exception
+     */
+    public static function getOrDefault($name, $default = null): mixed
```
Typehint `$name` and `$default` and then you can drop the docblocks.

jbrooksuk commented on src/Filament/Pages/Settings/ManageCachet.php:
```diff
@@ -98,6 +100,18 @@ public function form(Form $form): Form
                         ]),
+
+
+                    Forms\Components\Grid::make(2)
```
When enabling the API, we should make sure that `api_protected` is set to `false` again by default.

jbrooksuk reviewed (changes requested):
Couple of things we need to fix here.

icecoldPHP committed:
Fix when api disabled to reset api_protected to false
Fix add static types to method

jbrooksuk committed:
Merge branch 'main' into implement-235-toggle-api-auth-on-get

jbrooksuk commented:
Replaced in https://github.com/cachethq/core/pull/378
